main/src/main/scala/sbt/internal/server/BuildServerProtocol.scala

```diff
@@ -648,6 +687,33 @@ object BuildServerProtocol {
     )
   }
 
+  private def jvmEnvironmentItem(): Initialize[Task[JvmEnvironmentItem]] = Def.taskDyn {
+    val target = Keys.bspTargetIdentifier.value
+    val baseDir = Keys.baseDirectory.value.toURI().toString()
+    val jvmOptions = Keys.javaOptions.value.toVector
+    val env = envVars.value
+    val externalDependencyClasspath = Keys.externalDependencyClasspath.value
+
+    val internalDependencyClasspath = for {
+      (ref, configs) <- bspInternalDependencyConfigurations.value
+      config <- configs
+    } yield ref / config / Keys.classDirectory
```

Comment:
Is this a valid way to extract the classpath? I copied this from `scalacOptionsTask` (if yes then it can be probably refactored out)

Reply:
The easiest way to get the classpath is to use `Keys.fullClasspath.value`. The problem is that it triggers the compilation and it fails if there is any compiler error. That's why we could not use it in `scalacOption`. We want to import the build even if the compilation fails.

Do we want the `jvmRunEnvironement` to fail if the compilation fails or not? If the answer is no then this is the best way to extract the classpath that we have.

To refactor it you can use:
```scala
// does not trigger compilation
private def bspClasspathTask: Def.Initialize[Task[Vector[URI]]] = Def.taskDyn {
  val externalDependencyClasspath = Keys.externalDependencyClasspath.value

  val internalDependencyClasspath = for {
    (ref, configs) <- bspInternalDependencyConfigurations.value
    config <- configs
  } yield ref / config / Keys.classDirectory

  Def.task {
    val classpath = internalDependencyClasspath.join.value.distinct ++
      externalDependencyClasspath.map(_.data)
    classpath.map(_.toURI).toVector
  }
}
```

and call it like that:
```
Def.task {
  val target = Keys.bspTargetIdentifier.value
  val baseDir = Keys.baseDirectory.value.toURI().toString()
  val jvmOptions = Keys.javaOptions.value.toVector
  val env = envVars.value
  val classpath = bspClasspath.value
  JvmEnvironmentItem(target, classpath, jvmOptions, baseDir, env)
}
```

Reply:
> Do we want the `jvmRunEnvironement` to fail if the compilation fails or not?

According to the [spec](https://build-server-protocol.github.io/docs/extensions/jvm.html#test-environment-request) `The client should ask for it right before test/run execution, after all the targets are compiled.` so using `Keys.fullClasspath.value` might be a better way. The project should compile because otherwise, Metals won't be able to run it.

Speaking about protocol, you could look at https://github.com/build-server-protocol/build-server-protocol/pull/312 @adpi2, I made a small fixes to the docs.

Reply:
> So using Keys.fullClasspath.value might be a better way

It makes sense.

Thanks for pinging me on this build-server-protocol PR, I would have missed it otherwise.